diff --git a/colors.go b/colors.go
index 9520480..f7b1777 100644
--- a/colors.go
+++ b/colors.go
@@ -309,8 +309,6 @@ func (sm styleMap) get(f *file) tcell.Style {
 		key = "su"
 	case f.Mode()&os.ModeSetgid != 0:
 		key = "sg"
-	case f.Mode()&0111 != 0:
-		key = "ex"
 	}
 
 	if val, ok := sm[key]; ok {
diff --git a/icons.go b/icons.go
index ac4230f..fe31ffd 100644
--- a/icons.go
+++ b/icons.go
@@ -136,8 +136,6 @@ func (im iconMap) get(f *file) string {
 		key = "su"
 	case f.Mode()&os.ModeSetgid != 0:
 		key = "sg"
-	case f.Mode()&0111 != 0:
-		key = "ex"
 	}
 
 	if val, ok := im[key]; ok {
